dash/include/dash/pattern/internal/PatternArguments.h

```diff
@@ -121,72 +104,160 @@ class PatternArguments {
+   */
+  template<
+    int ArgcSize,
+    int ArgcDist,
+    int ArgcTeam,
+    int ArgcTeamSpec,
+    typename ... Args>
+  void parse(const DistributionSpec_t & ds, Args && ... args) {
+    static_assert(ArgcDist == 0, "Cannot mix DistributionSpec and inidividual"
+        "distributions in variadic pattern constructor!");
     DASH_LOG_TRACE("PatternArguments.check(distSpec)");
-    _argc_dist += NumDimensions;
-    _distspec   = ds;
+    _distspec = ds;
+    parse<ArgcSize, -1, ArgcTeam,
+    ArgcTeam>(std::forward<Args>(args)...);
   }
-  /// BlockPattern matching for up to NumDimensions optional parameters
-  /// specifying the distribution.
-  template<int count>
-  void check(const Distribution & ds) {
+
+  /*
+   * Match for up to \c NumDimensions optional parameters
+   * specifying the distribution.
+   */
+  template<
+    int ArgcSize,
+    int ArgcDist,
+    int ArgcTeam,
+    int ArgcTeamSpec,
+    typename ... Args>
+  void parse(const Distribution & ds, Args && ... args) {
+    static_assert(!(ArgcDist < 0), "Cannot mix DistributionSpec and "
+        "inidividual distributions in variadic pattern constructor!");
+    static_assert(ArgcDist < NumDimensions, "Number of distribution specifier"
+        " exceeds the number of dimensions in variadic pattern constructor!");
     DASH_LOG_TRACE("PatternArguments.check(dist)");
-    _argc_dist++;
-    dim_t dim = count - NumDimensions;
-    _distspec[dim] = ds;
-  }
-  /// Isolates first argument and calls the appropriate check() function
-  /// on each argument via recursion on the argument list.
-  template<int count, typename T, typename ... Args>
-  void check_recurse(T && t, Args && ... args) {
-    DASH_LOG_TRACE("PatternArguments.check(args) ",
-                   "count", count,
-                   "argc", sizeof...(Args));
-    check<count>(std::forward<T>(t));
-    if (sizeof...(Args) > 0) {
-      check_recurse<count + 1>(std::forward<Args>(args)...);
-    }
+    _distspec[ArgcDist] = ds;
+    parse<ArgcSize, ArgcDist+1, ArgcTeam,
+      ArgcTeam>(std::forward<Args>(args)...);
   }
-  /// Terminator function for recursive argument parsing
-  template<int count>
-  void check_recurse() {
+
+  /**
+   * Stop recursion when all arguments are processed.
+   */
+  template<
+    int ArgcSize,
+    int ArgcDist,
+    int ArgcTeam,
+    int ArgcTeamSpec>
+  void parse() {
+    static_assert(!(ArgcDist > 0 && ArgcDist != NumDimensions),
+        "Incomplete distribution specification in "
+        "variadic pattern constructor!");
+
+    static_assert(!(ArgcSize > 0 && ArgcSize != NumDimensions),
+        "Incomplete size specification in "
+        "variadic pattern constructor!");
+
+    /*
+     * TODO: we have no way to statically check whether a TeamSpec was created
+     *       using the team specified by the user.
+     */
+    if (ArgcTeam && !ArgcTeamSpec) {
+      _teamspec = TeamSpec_t(_distspec, *_team);
+    }
```

Comment:
The `dash::Team::All` is used by default if no team has been set, see `team()`. The assumption here is that it only makes sense to overwrite the default `TeamSpec` if a team was set. However, I realized that I am not sure about the relation between `DistSpec` and `TeamSpec` and whether we have to create the latter if the former was explicitely given bu the user. Is it even valid to specify both at the same time? @fuchsto can you shed some light?